inventory.lua

```diff
@@ -310,6 +314,100 @@ minetest.register_node("digilines:chest", {
 	end
 })
 
+if minetest.global_exists("tubelib") and minetest.global_exists("tubelib2") then
+	local speculative_pull = nil
+	local pull_succeeded = function(passed_speculative_pull)
+		if passed_speculative_pull.canceled then return end
+
+		send_message(passed_speculative_pull.pos, "ttake", passed_speculative_pull.taken,
+				passed_speculative_pull.index, nil, vector.multiply(passed_speculative_pull.dir, -1))
+		check_empty(passed_speculative_pull.pos)
+	end
+	local function tube_side(pos, side)
+		if side == "U" then return {x=0,y=-1,z=0}
+		elseif side == "D" then return {x=0,y=1,z=0}
+		end
+		local param2 = minetest.get_node(pos).param2
+		return vector.multiply(
+			minetest.facedir_to_dir(
+				tubelib2.side_to_dir(side, param2) - 1),
+			-1)
+	end
+	tubelib.register_node("digilines:chest", {}, {
+		on_pull_stack = function(pos, side, player_name)
+			local inv = minetest.get_meta(pos):get_inventory()
+			for i, stack in pairs(inv:get_list("main")) do
+				if not stack:is_empty() then
+					speculative_pull = {
+						canceled = false,
+						pos = pos,
+						taken = stack,
+						index = i,
+						dir = tube_side(pos, side)
+					}
+					minetest.after(0, pull_succeeded, speculative_pull)
+					inv:set_stack("main", i, nil)
+					return stack
+				end
+			end
+			return nil
+		end,
+		on_pull_item = function(pos, side, player_name)
+			local inv = minetest.get_meta(pos):get_inventory()
+			for i, stack in pairs(inv:get_list("main")) do
+				if not stack:is_empty() then
+					local taken = stack:take_item(1)
+					speculative_pull = {
+						canceled = false,
+						pos = pos,
+						taken = taken,
+						index = i,
+						dir = tube_side(pos, side)
+					}
+					minetest.after(0, pull_succeeded, speculative_pull)
+					inv:set_stack("main", i, stack)
+					return taken
+				end
+			end
+			return nil
+		end,
+		on_push_item = function(pos, side, item, player_name)
+			local dir_vec = tube_side(pos, side)
+			if tube_can_insert(pos, nil, item, dir_vec) then
+				tube_insert_object(pos, nil, item, dir_vec)
+				return true
+			end
+			return false
+		end,
+		on_unpull_item = function(pos, side, item, player_name)
+			local inv = minetest.get_meta(pos):get_inventory()
+			if inv:room_for_item("main", item) then
+				local existing_stack = inv:get_stack("main", speculative_pull.index)
+				local is_empty = existing_stack:is_empty()
+				if is_empty or (
+					existing_stack:get_name() == item:get_name() and
+					existing_stack:get_count()+item:get_count() <= existing_stack:get_stack_max())
```

Comment:
Possible simplification:
```Lua
local leftover = existing_stack:add_item(item)
if leftover:is_empty() then
	-- Stack fits
	inv:set_stack(.....)
	....
	return true
end
return false
```
I might have misunderstood the purpose of this function, though.

Reply:
That makes sense, didn't know about the leftover behaviour. What happens if `existing_stack` contains different items to `item`? Will `leftover` effectively be equivalent to `item`?

Reply:
yes. It'll just return the stuff that didn't fit. i.e. the entire stack.

Reply:
I haven't checked, probably should give it a test. It's pretty much impossible to trigger this logic anyways since it _should_ be an unreachable state

Reply:
@SmallJoker seems to be working as expected without error